### ISSUE TITLE
[patch] volume.detach() is broken

### ISSUE THREAD

**Issue by GoogleCodeExporter** [peterknife/boto#288]
```
Hi,

It looks like the volume.detach() call is broken. It uses self.volume_id
which is not initialized anywhere.

I don't quite see the purpose of passing volume id to detach call (check
against detaching a volume which has been already detached and attached
elsewhere?) but the patch below tries to obtain a volume id from
attach_data, falling back to None.

Why the fallback? Well, if a volume was not attached in the application,
then attached outside of the application, I might want to detach it from
the application (knowing that it is in fact attached) even if its status in
the application is 'available'.

Or, I haven't looked into it too much but if attaching a volume does not
immediately transition it into 'attached' state, then it is quite possible
for an application to have a volume in 'attaching' state which is in fact
'attached', and to attempt to detach such a volume.

In any event, the patch is below or you could remove instance_id argument
altogether since it's marked optional in the docs.

This issue happens on svn trunk.
```

Original issue reported on code.google.com by `oleg.pud...@gmail.com` on 27 Oct 2009 at 11:11

Attachments:
- [0002-Fixed-volume.detach-call.patch](https://storage.googleapis.com/google-code-attachments/boto/issue-288/comment-0/0002-Fixed-volume.detach-call.patch)
